### ISSUE TITLE
matchProjectTypes / excludeProjectTypes support in packageRules

### ISSUE THREAD

**Issue by Lctrs** [renovatebot/renovate#19207]
### What would you like Renovate to be able to do?

IMHO, it would be nice to be able to apply some package rules depending on the project type. For example, we would like to have a global config where we would pin all dependencies for binary project, but widen the supported ranges for libraries.

e.g. when using composer, you can define a [project type](https://getcomposer.org/doc/04-schema.md#type) in the composer.json file of the project. This field is optional, defaults to "library" and can be any string when present.

```json
{
    "name": "my/awesome-php-library",
    "type": "project",
    ...
}
```

At least with composer V1, this field is not copied to the composer.lock file.

I don't know if the project type can be determined for other project managers, but I guess it could be for some.

### If you have any ideas on how this should be implemented, please tell us here.

I have an embryo of implementation, but I have trouble getting the project type when applying package rules.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
Wondering if there's any managers other than `composer` which this could apply to. I think `projectType` and `matchProjectType` sounds like good naming though.

**Comment by Lctrs** [comment]
I think this could also apply to rust and cargo where projects created with `cargo init` could be of type `binary`, while projects created with ``cargo init --lib` could be of type `library`.

Also, you may want to match multiple project types, hence the plural.

**Comment by rarkins** [comment]
Yes, projectTypes can be a good idea, but then you need to decide if you want the match to be any or all?

**Comment by Lctrs** [comment]
Any.
